README.md
#250MilesCrossingPhila

Code and components developed to support the 250 Miles Crossing Philadelphia project

http://www.250miles.net/

##Milk Machine QGIS Python Plugin

The plugin can be found in the QGISPlugin folder. It is being developed in Python 2.7.5 64 bit, which is the native Python distribution for Quantum GIS (QGIS). The plugin was developed under QGIS 2.4, and has had minimal testing on QGIS 2.6. The plugin is compatible with Windows, Mac OSX, and Linux.

###QGIS

####Windows
To get QGIS For Windows, go to http://www.qgis.org/en/site/forusers/download.html and download the OSGeo4W Network Installer (64-bit), since it has the most comprehensive package list. Make sure to install the Python package `SciPy`.

####Linux
For Linux, get the latest stable build. 

`sudo apt-get install python-software-properties
sudo add-apt-repository ppa:ubuntugis/ubuntugis-unstable
sudo apt-get update
sudo apt-get install qgis python-qgis qgis-plugin-grass`

####Mac
http://www.kyngchaos.com/

###Install the Milk Machine Plugin

Copy the "MilkMachine" folder in the QGISPlugin folder to your ~\.qgis2\python\plugins\MilkMachine folder. This can be found in the
user directory under windows. In QGIS go to Plugins > Manage and Install Plugins... > Installed. Select the MilkMachine Plugin for
it to appear in the QGIS toolbar.

###Dependencies

###Credits/Funding Sources

University City Science Center
Wexford Science + Technology
Philadelphia Redevelopment Authority
Drexel University




QGISPlugin/ChangeLog.txt
Change Log
--------------------------------------------
Commit #15
Bug Fix: "Follow Behind" tab was disabled when all points were selected in active layer
Task: Enabled audio to be attached to kml/kmz when the end is beyond the length of the the GPS track
Task: Added datetime field. Populated on import with YYYY/MM/DD HH:MM:SS
Task: Changed all date and time references to point to the datetime field
Task: Enabled the audio sync to populate the "audio" field without being in an edit session
Task: Time tab. Edit selected time and before/after the selected time.
--------------------------------------------
Commit #16
Bug Fix: <TimeSpan> was defaulting to 5 seconds. Fixed to be 1 second.
Bug Fix: <Camera> heading was 180 degrees off what it should be.
Task: Added "relativeToModel" tag in the <altitude> for <Camera>. This means that the altitude will be relative to the 'altitude' tag in the <Model>, and that <Model> must be populated before attempting to use this setting.
Task: Added a yellow marker that indicates the selected point in an audio tour in QGIS
--------------------------------------------
Commit #17
Bug Fix: On windows, "Description" field formats as "Descriptio". On MAC, it format to "descriptio". MM handles both now.
--------------------------------------------
Commit #18
Bug Fix: Tab order on "Tour" tab.
Task: Check box for presence of yellow indicator over selected point in audio sync.
Bug Fix: If audio start time matches exactly with time of first GPS point: a) we get no notification window (window that gives audio and track info), b) MM does not generate a 'play-head', c) the audio does not play automatically when you click the play button
Task: "play" button is disabled when it is pressed. To enable it, press "stop" button.
--------------------------------------------
Commit #19
Bug Fix: Create <Camera> in the <document> before the <Tour> so that there are no strange camera actions.
Task: Make the "Follow Behind" applicable to a selection of points rather than automatically selecting the whole track
Task: Create a "smoother" field in "Follow Behind" tour so that the camera heading will take the mean angles of the points ahead of it.
Task: Create a "follow angle" field in "Follow Behind" tour. This field is the angle between the point and the camera. This enables the user to place the camera anywhere (0-360 degrees) around the model.
Task: Suggest a tilt angle based on the altitude and range values. This auto populates.
--------------------------------------------
Commit #20
Task: Allow for gaps in time in the <FlyTo> duration parameter. duration = time(n+1) - time(n)
Task: Pre-populate last known folder directory, for all browsing dialog boxes.
Task: Make the audio import file path the same as the export audio path.
--------------------------------------------
Commit #21
Bug Fix: The audio delay calculation was not firing since I was only inheriting the path from the import, rather than doind the calculation on the export tab.
--------------------------------------------
Commit #22
Task: The MM sync tab now behaves like GE. If there is a time gap between points, MM will continue to add time in the "Audio Clock Time" and "Duration Into Audio" time boxes. However, the FID box will stay static as well as the selected point on the map. These will again change, once the gps point timestamp matches the audio clock.
--------------------------------------------
Commit #22
Bug Fix: The 'camera' field exceeds the maximum character limit of 254 sometimes. This means that QGIS cuts of extra characters when saving as a shapefile. However, during and edit session the field will hold the extra characters. This allows a safe export in MM. There is no error in QGIS for the extra characters. The solution is to code all of the parameters. This brings the camera field down to about 180 characters.
Bug Fix: The 'tilt' field needs only to autopopulate when it is empty, not when the user enters a value. This stops it from overwriting the user values.
--------------------------------------------
Commit #23
Bug Fix: The custom camera function is was not fully obfuscating the camera dictionary. Result was writing the full dictionary, and it was still getting cut off by the 254 character limit.
--------------------------------------------
Commit #24
Task: Create a <LookAt> tab in the Tour tab of Milk Machine. The <LookAt> tab has two modes; "custom" and "circle around". Both automatically calculate latitude and longitude. In custom mode, the user can select 1 or more points and apply a custom <LookAt> to each. If multiple points are selected, lat/long for each <LookAt> will be the lat/long of that point/model. In "circle around" mode, 1 or more points can also be selected. If multiple points are selected, MM calculates the geometric centroid of the selected points for the <LookAt> lat/long. If only 1 point is selected, it's lat/long will be the lat/long for the <LookAt>. "circle around" also has some parameter for customizing the number of rotations, starting heading, and duration.
--------------------------------------------
Commit #26
Task: Change all text to Tahoma
Task: Update the "sync" button info box to indicate the row number.
Task: Filtering added with 3 point filters; linear, quadratic, and moving average.
--------------------------------------------
Commit #27
Bug Fix: Size policy changes for labels and buttons. Hoping that "fixed" "fixed" works
Bug Fix: The matplotlib library is not that same as the one on Windows, so no plotting!!. Only works on windows.
Bug Fix: numpy on Mac is an older version, so I can't call scipy/numpy as pylab. Call as numpy.
--------------------------------------------
Commit #28
Task: Upgrade the "circle around" to match the duration of the selected points. "duration" option is disabled, but autopopulates based on # selected features.
Still scratching my head on the font sizing. I may need to do some manual resizing if the user is on a Mac. Maybe.

Web/lib/js/grids.js
/**
 * @author OAXMCGI
 */
var grids = {"features":[
                    // Grid Cell 1
                    {"attributes":{"FID":1,"Id":1}, "tracks":[
                        {"Id":1, "type":"object", "movement":"stationary", "title": "General Title #1", "description": "Fridge in Uzbekistan", "complete":false}
                    ]},
                    // Grid Cell 2
                    {"attributes":{"FID":2,"Id":2}, "tracks":[
                        {"Id":2,"type":"human", "movement": "mobile", "title": "General Title",  "description":"Kelly de Kapster", "complete":false}
                    ]},
                    // Grid Cell 3
                    {"attributes":{"FID":3,"Id":3}, "tracks": [
                        {"Id":3,"type":"object", "movement": "stationary", "title": "General Title", "description":"Powerline", "complete":false}
                    ]},
                    // Grid Cell 4
                    {"attributes":{"FID":4,"Id":4},"tracks":[
                        {"Id":4,"type":"object", "movement": "stationary", "title": "General Title", "description":"No Descsription", "complete":false}
                    ]},
                    // Grid Cell 5
                    {"attributes":{"FID":5,"Id":5},"tracks":[
                        {"Id":6, "type":"human", "movement": "mobile", "title": "General Title", "description":"Angela's brother Alex Duncan", "complete":false}
                    ]},
                    // Grid Cell 6
                    {"attributes":{"FID":6,"Id":6}, "tracks":[
                        {"Id":7,"type":"human", "movement": "mobile", "title": "General Title", "description":"Ursula", "complete":false}
                    ]},
                    // Grid Cell 7
                    {"attributes":{"FID":7,"Id":7}, "tracks":[
                        {"Id":8,"type":"object", "movement": "mobile", "title": "General Title", "description":"Dog in Pennypack Park", "complete":false}
                    ]},
                    // Grid Cell 8
                    {"attributes":{"FID":8,"Id":8}, "tracks":[
                        {"Id":9,"type":"object", "movement": "mobile", "title": "General Title", "description":"Airplane from local airport", "complete":false}
                    ]},
                    // Grid Cell 9
                    {"attributes":{"FID":9,"Id":9}, "tracks":[
                        {"Id":10,"type":"object", "movement": "mobile", "title": "General Title", "description":"de hoge hakken van een dame", "complete":false}
                    ]},
                    // Grid Cell 10
                    {"attributes":{"FID":10,"Id":10}, "tracks":[
                        {"Id":11,"type":"human", "movement": "mobile", "title": "General Title", "description":"Richard Ramson a.k.a. RAM", "complete":false}
                    ]},
                    // Grid Cell 11
                    {"attributes":{"FID":11,"Id":11},"tracks": [
                        {"Id":12,"type":"human", "movement": "mobile", "title": "The biker", "description":"Riding through a city on a bicycle is an interesting experience. In fact you are in a middle position, faster then pedestrians, slower then cars (or other motorized vehicles). You could look at that as a no man’s land, no party is genuinely interested in you, unless you’re in their way.", "complete":true, "filename":"Akh-5min.kmz"}
                    ]},
                    // Grid Cell 12
                    {"attributes":{"FID":12,"Id":12},"tracks":[
                        {"Id":13,"type":"object", "movement": "mobile", "title": "General Title", "description":"Tacony-Palmyra Bridge Recording", "complete":false},
                        {"Id":14,"type":"object", "movement": "stationary", "title": "General Title", "description":"Revolutionary Recovery", "complete":false},
                        {"Id":15,"type":"object", "movement": "stationary", "title": "General Title", "description":"Priest statue", "complete":false}
                    ]},
                    // Grid Cell 13
                    {"attributes":{"FID":13,"Id":13}, "tracks":[
                        {"Id":16,"type":"human", "movement": "mobile", "title": "General Title", "description":"Former inmate walking along State Road corridor", "complete":false},
                        {"Id":17,"type":"object", "movement": "mobile", "title": "General Title", "description":"Sound recording made in car", "complete":false}
                    ]},
                    // Grid Cell 14
                    {"attributes":{"FID":14,"Id":14},"tracks":[
                        {"Id":18,"type":"object", "movement": "mobile", "title": "The mail bag", "description":"As a mail bag you cooperate closely with your carrier, the mail man. Your duty is to get personal envelopes to the rightful owners and you will do everything in your power to do this job well. You will never get the credit, but you will neither ever be blamed.", "complete":true, "filename":"Mailman-5min.kmz"}
                    ]},
                    // Grid Cell 15
                    {"attributes":{"FID":15,"Id":15},"tracks":[
                        {"Id":19,"type":"human", "movement": "mobile", "title": "General Title", "description":"protagonist walks up 52nd Street from Apple Storage Building (crossing from Block 19)", "complete":false},
                        {"Id":20,"type":"object", "movement": "mobile", "title": "General Title", "description":"trolley 15", "complete":false}
                    ]},
                    // Grid Cell 16
                    {"attributes":{"FID":16,"Id":16},"tracks":[
                        {"Id":21,"type":"human", "movement": "mobile", "title": "", "description":"Fireworks", "complete":false},
                        {"Id":22,"type":"object", "movement": "mobile", "title": "One eyed Dusty", "description":"A humming rider on your back, the sound of your hooves beneath you, passing cars left of you, just one eye, your right eye! The ultimate sound of echo’s that keeps you aware of the space on your left side. The rider has a steady hand.", "complete":true, "filename":"Horse.kmz" },
                        {"Id":23,"type":"object", "movement": "mobile", "title": "General Title", "description":"Dog", "complete":false}
                    ]},
                    // Grid Cell 17
                    {"attributes":{"FID":17,"Id":17},"tracks":[
                        {"Id":24,"type":"human", "movement": "mobile", "title": "General Title", "description":"Walter West", "complete":false},
                        {"Id":25,"type":"human", "movement": "mobile", "title": "General Title", "description":"Bill Callahan", "complete":false}
                    ]},
                    // Grid Cell 18
                    {"attributes":{"FID":18,"Id":18},"tracks":[
                        {"Id":26,"type":"object", "movement": "mobile", "title": "General Title", "description":"boat on river", "complete":false}
                    ]},
                    // Grid Cell 19
                    {"attributes":{"FID":19,"Id":19},"tracks":[
                        {"Id":27,"type":"human", "movement": "mobile", "title": "General Title", "description":"Brujo de la Mancha", "complete":false},
                        {"Id":28,"type":"human", "movement": "mobile", "title": "General Title", "description":"protagonist walks up 52nd Street from Apple Storage Building (crossing from Block 19)", "complete":false}
                    ]},
                    // Grid Cell 20
                    {"attributes":{"FID":20,"Id":20},"tracks":[
                        {"Id":29,"type":"human", "movement": "mobile", "title": "General Title", "description":"Austin Seraphin - blind echo locator", "complete":false},
                        {"Id":30,"type":"human", "movement": "mobile", "title": "General Title", "description":"Melodi in Chinatown", "complete":false}
                    ]},
                    // Grid Cell 21
                    {"attributes":{"FID":21,"Id":21},"tracks":[
                        {"Id":31,"type":"object", "movement": "stationary", "title": "General Title", "description":"Pier/Jetty", "complete":false}
                    ]},
                    // Grid Cell 22
                    {"attributes":{"FID":22,"Id":22},"tracks":[
                        {"Id":32,"type":"object", "movement": "stationary", "title": "General Title", "description":"Viaduct", "complete":false},
                        {"Id":33,"type":"human", "movement": "mobile", "title": "General Title", "description":"2 girls in car read Walt Whitman poem", "complete":false}
                    ]},
                    // Grid Cell 23
                    {"attributes":{"FID":23,"Id":23},"tracks":[
                        {"Id":34,"type":"object", "movement": "mobile", "title": "General Title", "description":"forklift", "complete":false},
                        {"Id":35,"type":"human", "movement": "mobile", "title": "General Title", "description":"2 girls in car read Walt Whitman poem", "complete":false}
                    ]},
                    // Grid Cell 24
                    {"attributes":{"FID":24,"Id":24},"tracks":[
                        {"Id":36,"type":"human", "movement": "mobile", "title": "General Title", "description":"2 girls in car read Walt Whitman poem", "complete":false}
                    ]}
                ]
            },
    bgImages = [
                {image : 'lib/images/1.jpg'},
                {image : 'lib/images/2.jpg'},  
                {image : 'lib/images/3.jpg'},
                {image : 'lib/images/4.jpg'},
                {image : 'lib/images/5.jpg'},
                {image : 'lib/images/6.jpg'},
                {image : 'lib/images/7.jpg'},
                {image : 'lib/images/8.jpg'},
                {image : 'lib/images/9.jpg'},
                {image : 'lib/images/10.jpg'},
                {image : 'lib/images/11.jpg'},
                {image : 'lib/images/12.jpg'},
                {image : 'lib/images/13.jpg'},
                {image : 'lib/images/14.jpg'},
                {image : 'lib/images/15.jpg'},
                {image : 'lib/images/16.jpg'},
                {image : 'lib/images/17.jpg'}
        ];
            
